File: backend/utils/smsSecuritySpec.js
/**
 * smsSecuritySpec.js — Canonical Cryptographic Contract (Server Mirror)
 * ======================================================================
 *
 * PURPOSE
 * -------
 * This file is the SERVER-SIDE MIRROR of the canonical cryptographic and
 * data-integrity contract for the SMS payment pipeline.
 *
 * The authoritative Android source is:
 *   app/src/main/java/online/paychek/app/utils/SmsSecuritySpec.kt
 *
 * Both files MUST remain in sync. Any change here must be reflected there,
 * and vice versa.
 *
 * Server implementations referenced by this spec:
 *   - backend/utils/verifyHmac.js       (HMAC verification)
 *   - backend/controllers/paymentController.js  (sha256 helper + ingest logic)
 *   - backend/schema.sql                (dedupe_key formula)
 *
 * ======================================================================
 * CONTRACT SECTION 1 — rawBody Immutability
 * ======================================================================
 *
 * RULE: rawBody is the IMMUTABLE byte-for-byte SMS body string as received
 *       from the Android client. It must NEVER be modified before hashing.
 *
 * ALLOWED on server:
 *   const { rawBody } = req.body;   // receive and store as-is
 *   verifyHmac(rawBody, signature, secretKey);  // hash the exact string
 *
 * FORBIDDEN (will silently break HMAC + duplicate detection):
 *   rawBody.trim()
 *   rawBody.toLowerCase()
 *   rawBody.replace(...)
 *   rawBody + anything
 *
 * WHY: Both HMAC-SHA256 and SHA-256(rawBody) are computed on this string.
 *      Any mutation causes:
 *        - Server HMAC != Android HMAC  -> payload rejected as tampered
 *        - Server SHA256 != Android SHA256 -> deduplication ghost entries
 *
 * ======================================================================
 * CONTRACT SECTION 2 — HMAC-SHA256 Specification
 * ======================================================================
 *
 * ALGORITHM:  HMAC-SHA256
 * KEY:        per-user secret stored in users table (secretKey column)
 * INPUT:      UTF-8 string rawBody (immutable, see Section 1)
 * OUTPUT:     lowercase hexadecimal string (64 chars)
 *
 * Server implementation (verifyHmac.js):
 *   crypto.createHmac('sha256', secretKey)
 *          .update(rawBody, 'utf8')
 *          .digest('hex')
 *
 * Android implementation (HmacHelper.generate):
 *   Mac.getInstance("HmacSHA256")
 *   mac.doFinal(rawBody.toByteArray(UTF_8))
 *   -> joinToString("") { "%02x".format(it) }
 *
 * VERIFICATION: Use crypto.timingSafeEqual() — NEVER use === for comparison.
 *
 * ======================================================================
 * CONTRACT SECTION 3 — SHA-256 rawBodyHash Specification
 * ======================================================================
 *
 * PURPOSE:     Deduplication key component in sms_history table.
 *
 * ALGORITHM:  SHA-256
 * INPUT:      UTF-8 string rawBody (immutable, see Section 1)
 * OUTPUT:     lowercase hexadecimal string (64 chars)
 *
 * Server implementation (paymentController.js sha256 helper):
 *   crypto.createHash('sha256').update(data, 'utf8').digest('hex')
 *
 * Android implementation (HmacHelper.sha256Hex):
 *   MessageDigest.getInstance("SHA-256")
 *   digest(rawBody.toByteArray(UTF_8))
 *   -> joinToString("") { "%02x".format(it) }
 *
 * Server deduplication key formula (schema.sql):
 *   dedupe_key = CONCAT(sms_timestamp, '|', sender, '|', SHA256(raw_body))
 *
 * ======================================================================
 * CONTRACT SECTION 4 — Offline Queue Retry Policy
 * ======================================================================
 *
 * MAX_RETRIES:   10  (enforced in Android DAO query)
 * BACKOFF:       Exponential (enforced in SmsReceiver.calculateNextRetryMs)
 *   Attempt 0  -> retry in 30 seconds
 *   Attempt 1  -> retry in 2 minutes
 *   Attempt 2  -> retry in 10 minutes
 *   Attempt 3  -> retry in 1 hour
 *   Attempt 4+ -> retry in 6 hours (cap)
 *
 * SERVER RESPONSE CODES (must match what Android client expects):
 *   200 / 201 -> success      -> Android: markAsSynced()
 *   401       -> HMAC fail    -> Android: markRetryFailed() (may fix itself)
 *   409       -> duplicate    -> Android: markAsSynced() (already on server)
 *   422       -> parse fail   -> Android: markPermanentlyFailed() — DO NOT change to 4xx else
 *   5xx       -> server error -> Android: markRetryFailed() with backoff
 *
 * CRITICAL: HTTP 422 is the PERMANENT FAILURE signal.
 *   If the server changes this code, Android retry logic breaks.
 *   The Android client will retry 422 indefinitely if the code changes.
 *
 * ======================================================================
 * CONTRACT SECTION 5 — Ingest Request Fields
 * ======================================================================
 *
 * Android sends to POST /api/payments/sms-ingest:
 *   rawBody        {string}  exact SMS body — immutable (Section 1)
 *   hmacSignature  {string}  HMAC-SHA256 hex (Section 2)
 *   smsTimestamp   {number}  Unix ms — from Android SmsMessage.timestampMillis
 *   simSlot        {number}  1 or 2 — physical SIM slot index
 *   simNumber      {string?} phone number of the SIM (may be null)
 *   isOfflineSync  {boolean} true when replaying from Room offline queue
 *
 * Server MUST process in this order:
 *   1. Verify HMAC  (reject 401 on mismatch — do not proceed)
 *   2. Parse rawBody using stored provider regex template
 *   3. Compute SHA256(rawBody) to build dedupe_key
 *   4. Check sms_history for dedupe_key collision  (return 409 if exists)
 *   5. Insert into sms_history
 *   6. Return 422 if step 2 fails  (parse failure — do not retry)
 *
 * ======================================================================
 * CHANGE POLICY
 * ======================================================================
 *
 * Any change to the above contracts requires ALL of the following:
 *   [ ] Update this file (smsSecuritySpec.js)
 *   [ ] Update SmsSecuritySpec.kt (Android mirror)
 *   [ ] Update verifyHmac.js (server HMAC implementation)
 *   [ ] Update paymentController.js sha256() helper if needed
 *   [ ] Update HmacHelper.kt (Android HMAC + SHA256 implementation)
 *   [ ] Increment API version if encoding or algorithm changes
 *   [ ] Cross-verify: compute hash of a known rawBody on both sides
 *       and confirm the outputs are identical before deploying
 *
 * ======================================================================
 */

// -------------------------------------------------------------------------
// Section 2 constants — HMAC
// -------------------------------------------------------------------------
const HMAC_ALGORITHM      = 'sha256';         // crypto.createHmac arg
const HMAC_ENCODING       = 'utf8';           // .update() encoding arg
const HMAC_OUTPUT_FORMAT  = 'hex';            // .digest() format (lowercase)

// -------------------------------------------------------------------------
// Section 3 constants — SHA-256 rawBodyHash
// -------------------------------------------------------------------------
const HASH_ALGORITHM      = 'sha256';         // crypto.createHash arg
const HASH_ENCODING       = 'utf8';           // .update() encoding arg
const HASH_OUTPUT_FORMAT  = 'hex';            // .digest() format (lowercase)

// -------------------------------------------------------------------------
// Section 4 constants — HTTP response codes Android client expects
// -------------------------------------------------------------------------
const HTTP_SUCCESS        = 200;
const HTTP_CREATED        = 201;
const HTTP_DUPLICATE      = 409;  // already exists — Android marks as synced
const HTTP_HMAC_FAIL      = 401;  // HMAC mismatch — reject without processing
const HTTP_PARSE_FAIL     = 422;  // SMS unparseable — Android PERMANENT FAIL (no retry)

// -------------------------------------------------------------------------
// Section 4 constants — Retry schedule (for documentation; enforced on Android)
// -------------------------------------------------------------------------
const RETRY_MAX_ATTEMPTS  = 10;
const RETRY_DELAYS_MS     = [30_000, 120_000, 600_000, 3_600_000, 21_600_000]; // cap at index 4

// -------------------------------------------------------------------------
// Section 1 — rawBody validation helper
// -------------------------------------------------------------------------
/**
 * Validates that rawBody has not been mutated before hashing.
 * Returns true if safe to proceed; throws if rawBody is empty/null.
 *
 * @param {string} rawBody
 * @returns {boolean}
 */
function assertRawBodyIntegrity(rawBody) {
    if (!rawBody || typeof rawBody !== 'string' || rawBody.length === 0) {
        throw new Error(
            '[SmsSecuritySpec] rawBody integrity violation: ' +
            'rawBody is null, undefined, or empty. ' +
            'This indicates rawBody was modified before reaching this point. ' +
            'See smsSecuritySpec.js Section 1.'
        );
    }
    return true;
}

module.exports = {
    HMAC_ALGORITHM,
    HMAC_ENCODING,
    HMAC_OUTPUT_FORMAT,
    HASH_ALGORITHM,
    HASH_ENCODING,
    HASH_OUTPUT_FORMAT,
    HTTP_SUCCESS,
    HTTP_CREATED,
    HTTP_DUPLICATE,
    HTTP_HMAC_FAIL,
    HTTP_PARSE_FAIL,
    RETRY_MAX_ATTEMPTS,
    RETRY_DELAYS_MS,
    assertRawBodyIntegrity,
};

File: app/app/src/main/java/online/paychek/app/utils/SmsSecuritySpec.kt
package online.paychek.app.utils

/**
 * SmsSecuritySpec — Canonical Cryptographic Contract for the Payment Checker System
 * ==================================================================================
 *
 * PURPOSE
 * -------
 * This file is the SINGLE SOURCE OF TRUTH for all cryptographic and
 * data-integrity rules in the SMS payment pipeline.
 *
 * Every developer, code reviewer, or AI assistant touching crypto-related
 * code MUST consult this file first. Any change to these rules MUST be
 * mirrored in:
 *   - Android: HmacHelper.kt (implementation)
 *   - Server:  backend/utils/smsSecuritySpec.js (mirror contract)
 *   - Server:  backend/utils/verifyHmac.js (implementation)
 *
 * ==================================================================================
 * CONTRACT SECTION 1 — rawBody Immutability
 * ==================================================================================
 *
 * RULE: rawBody is the IMMUTABLE byte-for-byte SMS body string as received from
 *       android.telephony.SmsMessage.getMessageBody()
 *
 * ALLOWED:
 *   val body = msg.messageBody ?: continue   <- assign once, from OS
 *
 * FORBIDDEN (will silently break HMAC + duplicate detection):
 *   body.trim()
 *   body.lowercase()
 *   body.replace(...)
 *   body + anything
 *   anything + body
 *   body.normalize()
 *
 * WHY: HMAC-SHA256 and SHA-256(rawBody) are both computed from this exact string.
 *      Any byte-level change causes:
 *        - Android HMAC != Server HMAC  -> server rejects as tampered
 *        - Android SHA256 != Server SHA256 -> deduplication breaks silently
 *
 * ==================================================================================
 * CONTRACT SECTION 2 — HMAC-SHA256 Specification
 * ==================================================================================
 *
 * ALGORITHM:  HMAC-SHA256
 * KEY:        per-user 64-char hex secret, stored in EncryptedSharedPreferences
 *             key name: SmsReceiver.KEY_HMAC_SECRET ("pcu_hmac_secret_key_v2")
 * INPUT:      UTF-8 bytes of rawBody (immutable, see Section 1)
 * KEY INPUT:  UTF-8 bytes of secretKey
 * OUTPUT:     lowercase hexadecimal string (64 chars)
 *
 * Android (canonical implementation -> HmacHelper.generate):
 *   Mac.getInstance("HmacSHA256")
 *   SecretKeySpec(secretKey.toByteArray(UTF_8), "HmacSHA256")
 *   mac.doFinal(rawBody.toByteArray(UTF_8))
 *   -> joinToString("") { "%02x".format(it) }
 *
 * Server (canonical implementation -> verifyHmac.js):
 *   crypto.createHmac('sha256', secretKey)
 *        .update(rawBody, 'utf8')
 *        .digest('hex')
 *
 * VERIFICATION: Server uses crypto.timingSafeEqual() to prevent timing attacks.
 *
 * ==================================================================================
 * CONTRACT SECTION 3 — SHA-256 rawBodyHash Specification
 * ==================================================================================
 *
 * PURPOSE:     Deduplication key component in sms_history table on the server.
 *              Stored in PendingSmsEntity.rawBodyHash (Room DB).
 *
 * ALGORITHM:  SHA-256
 * INPUT:      UTF-8 bytes of rawBody (immutable, see Section 1)
 * OUTPUT:     lowercase hexadecimal string (64 chars)
 *
 * Android (canonical implementation -> HmacHelper.sha256Hex):
 *   MessageDigest.getInstance("SHA-256")
 *   digest(rawBody.toByteArray(UTF_8))
 *   -> joinToString("") { "%02x".format(it) }
 *
 * Server (inline in paymentController.js sha256 helper):
 *   crypto.createHash('sha256').update(data, 'utf8').digest('hex')
 *
 * Server deduplication key formula (schema.sql):
 *   dedupe_key = CONCAT(sms_timestamp, '|', sender, '|', SHA256(raw_body))
 *
 * ==================================================================================
 * CONTRACT SECTION 4 — Offline Queue Retry Policy
 * ==================================================================================
 *
 * MAX_RETRIES:   10  (enforced in PendingSmsDao query: retryCount < 10)
 * BACKOFF:       Exponential (see SmsReceiver.calculateNextRetryMs)
 *   Attempt 0  -> retry in 30 seconds
 *   Attempt 1  -> retry in 2 minutes
 *   Attempt 2  -> retry in 10 minutes
 *   Attempt 3  -> retry in 1 hour
 *   Attempt 4+ -> retry in 6 hours (cap)
 *
 * PERMANENT FAILURE: HTTP 422 (SMS_PARSE_FAILED)
 *   -> markPermanentlyFailed() called immediately
 *   -> Item removed from active retry queue
 *   -> Reason: server HMAC passed but SMS body unparseable; retrying is futile
 *
 * CLEANUP: Synced entries older than 7 days are deleted from Room DB.
 *
 * ==================================================================================
 * CONTRACT SECTION 5 — Server-Side Ingest Rules
 * ==================================================================================
 *
 * The Android client sends the following fields to POST /api/payments/sms-ingest:
 *   rawBody        <- exact immutable SMS body (Section 1)
 *   hmacSignature  <- HMAC-SHA256(rawBody, secretKey) (Section 2)
 *   smsTimestamp   <- Unix ms timestamp from SmsMessage.timestampMillis
 *   simSlot        <- 1 or 2 (physical SIM slot)
 *   simNumber      <- phone number of the SIM (may be null)
 *   isOfflineSync  <- true when replaying from Room queue
 *
 * The server MUST:
 *   1. Verify HMAC first (reject 401 if mismatch)
 *   2. Parse rawBody using stored regex templates
 *   3. Compute SHA256(rawBody) for dedupe_key
 *   4. Return 422 if parsing fails (client must NOT retry)
 *   5. Return 409 if dedupe_key already exists (client marks as synced)
 *   6. Return 200/201 on success
 *
 * ==================================================================================
 * CHANGE POLICY
 * ==================================================================================
 *
 * Any change to the above contracts requires:
 *   [ ] Update this file (SmsSecuritySpec.kt)
 *   [ ] Update smsSecuritySpec.js (backend mirror)
 *   [ ] Update HmacHelper.kt (Android implementation)
 *   [ ] Update verifyHmac.js (server implementation)
 *   [ ] Update paymentController.js sha256() helper if needed
 *   [ ] Increment API version if breaking (e.g. encoding change)
 *   [ ] Test with a known rawBody to verify Android hash == Server hash
 *
 * ==================================================================================
 */
object SmsSecuritySpec {

    // -------------------------------------------------------------------------
    // Section 2 constants — HMAC
    // -------------------------------------------------------------------------
    const val HMAC_ALGORITHM          = "HmacSHA256"
    const val HMAC_KEY_PREFS_KEY      = "pcu_hmac_secret_key_v2"
    const val HMAC_OUTPUT_FORMAT      = "lowercase_hex"
    const val HMAC_ENCODING           = "UTF-8"
    const val HMAC_MIN_KEY_LENGTH     = 32

    // -------------------------------------------------------------------------
    // Section 3 constants — SHA-256 rawBodyHash
    // -------------------------------------------------------------------------
    const val HASH_ALGORITHM          = "SHA-256"
    const val HASH_OUTPUT_FORMAT      = "lowercase_hex"
    const val HASH_ENCODING           = "UTF-8"

    // -------------------------------------------------------------------------
    // Section 4 constants — Retry policy
    // -------------------------------------------------------------------------
    const val RETRY_MAX_ATTEMPTS      = 10
    const val RETRY_DELAY_0_MS        = 30_000L       // 30 seconds
    const val RETRY_DELAY_1_MS        = 120_000L      // 2 minutes
    const val RETRY_DELAY_2_MS        = 600_000L      // 10 minutes
    const val RETRY_DELAY_3_MS        = 3_600_000L    // 1 hour
    const val RETRY_DELAY_CAP_MS      = 21_600_000L   // 6 hours (cap)
    const val SYNC_CLEANUP_AGE_MS     = 7L * 24 * 60 * 60 * 1_000  // 7 days

    // HTTP response codes from server
    const val HTTP_HMAC_FAIL          = 401
    const val HTTP_PARSE_FAIL         = 422  // permanent — do not retry
    const val HTTP_DUPLICATE          = 409  // already exists — mark as synced

    // -------------------------------------------------------------------------
    // Section 1 — rawBody immutability check (runtime assertion)
    // Use this in any code path that receives a rawBody before hashing.
    // -------------------------------------------------------------------------
    /**
     * Asserts that rawBody is not blank.
     * Returns true if safe to proceed; false if the item should be aborted.
     *
     * DOES NOT modify rawBody. The caller must abort if this returns false.
     */
    fun isRawBodyValid(rawBody: String): Boolean = rawBody.isNotBlank()
}
